pkg/tso/global_allocator.go

```diff
@@ -651,10 +677,57 @@ func (gta *GlobalTSOAllocator) campaignLeader() {
 			log.Info("exit leader campaign",
 				logutil.CondUint32("keyspace-group-id", gta.getGroupID(), gta.getGroupID() > 0))
 			return
+		case <-exitPrimary:
+			log.Info("no longer be primary because primary have been updated, the TSO primary will step down")
+			return
 		}
 	}
 }
 
+// keepExpectedPrimaryAlive keeps the expected primary alive.
+// We use lease to keep `expected primary` healthy.
+// ONLY reset by the following conditions:
+// - changed by`/ms/primary/transfer` API.
+// - server closed.
+func (gta *GlobalTSOAllocator) keepExpectedPrimaryAlive(ctx context.Context) (*election.Leadership, int64, error) {
+	const purpose = "tso-primary-watch"
+	cli := gta.member.Client()
+	newLease := election.NewLease(cli, purpose)
+	if err := newLease.Grant(gta.am.leaderLease); err != nil {
+		return nil, 0, err
+	}
+
+	revision, err := mcsutils.MarkExpectedPrimaryFlag(cli, gta.member.GetLeaderPath(), gta.member.MemberValue(),
+		newLease.ID.Load().(clientv3.LeaseID))
+	if err != nil {
+		log.Error("mark expected primary error", errs.ZapError(err))
+		return nil, 0, err
+	}
+	// Keep alive the current primary leadership to indicate that the server is still alive.
+	// Watch the expected primary path to check whether the expected primary has changed.
+	expectedLease := election.NewLeadership(cli, mcsutils.ExpectedPrimaryPath(gta.member.GetLeaderPath()), purpose)
+	expectedLease.SetLease(newLease)
+	expectedLease.Keep(ctx)
+	return expectedLease, revision, nil
+}
+
+// primaryWatch watches `/ms/primary/transfer` API whether changed the expected primary.
+func (gta *GlobalTSOAllocator) expectedPrimaryWatch(ctx context.Context, expectedLease *election.Leadership, revision int64, exitPrimary chan struct{}) {
```

Comment:
It seems that we will always call a goroutine to run `expectedPrimaryWatch` after `keepExpectedPrimaryAlive`. What about merging them into a single call?